### PR TITLE
fix: create separate Line Items for Batches and Package Tags

### PR DESCRIPTION
**Pull request by MyuddinKhatri**
Ref: [Asana](https://app.asana.com/0/1192407896547052/1200099124039411)

### CODE REVIEW

File: erpnext/stock/doctype/stock_reconciliation/stock_reconciliation.py [hunk 1]
```diff
@@ -465,20 +465,59 @@ def get_items(warehouse, posting_date, posting_time, company):
...
-				"item_name": d[1],
-				"valuation_rate": stock_bal[1],
-				"current_qty": stock_bal[0],
-				"current_valuation_rate": stock_bal[1]
-			})
+			package_tags = frappe.db.sql("""
+				Select distinct sle.package_tag, sle.actual_qty, sle.valuation_rate
+				from `tabStock Ledger Entry` sle
+				where sle.item_code = %s and sle.warehouse = %s and sle.package_tag != ''
+			""",(d[0], d[2]), as_dict=1)
```
Comment (hrwX) [inline erpnext/stock/doctype/stock_reconciliation/stock_reconciliation.py]: Use ORM

File: erpnext/stock/doctype/stock_reconciliation/stock_reconciliation.py [hunk 2]
```diff
@@ -465,20 +465,59 @@ def get_items(warehouse, posting_date, posting_time, company):
...
+
+			batch_nos = frappe.db.sql("""
+				Select distinct sle.batch_no, sle.actual_qty, sle.valuation_rate
+				from `tabStock Ledger Entry` sle
+				where sle.item_code = %s and sle.warehouse = %s and sle.batch_no != ''
+			""",(d[0], d[2]), as_dict=1)
```
Comment (hrwX) [inline erpnext/stock/doctype/stock_reconciliation/stock_reconciliation.py]: Use ORM

File: erpnext/stock/doctype/stock_reconciliation/stock_reconciliation.py [hunk 3]
```diff
@@ -465,20 +465,59 @@ def get_items(warehouse, posting_date, posting_time, company):
...
+						res.append({
+							"item_code": d[0],
+							"warehouse": d[2],
+							"batch_no": batch_no.batch_no,
+							"qty": batch_no.actual_qty,
+							"item_name": d[1],
+							"valuation_rate": batch_no.valuation_rate,
+							"current_qty": stock_bal[0],
+							"current_valuation_rate": stock_bal[1]
+						})
```
Comment (hrwX) [inline erpnext/stock/doctype/stock_reconciliation/stock_reconciliation.py]:
```suggestion
			if package_tags:
				for package_tag in package_tags:
					res.append({
						"item_code": d[0],
						"warehouse": d[2],
					        "package_tag": package_tag.package_tag,
					        "qty": package_tag.actual_qty,
					        "item_name": d[1],
					        "valuation_rate": package_tag.valuation_rate,
					        "current_qty": stock_bal[0],
					        "current_valuation_rate": stock_bal[1]
				        })

		          if batch_nos:
			          for batch_no in batch_nos:
				          res.append({
					          "item_code": d[0],
					          "warehouse": d[2],
					          "batch_no": batch_no.batch_no,
					          "qty": batch_no.actual_qty,
					          "item_name": d[1],
					          "valuation_rate": batch_no.valuation_rate,
					          "current_qty": stock_bal[0],
					          "current_valuation_rate": stock_bal[1]
				          })
```

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/stock/doctype/stock_reconciliation/stock_reconciliation.py [hunk 4]
```diff
@@ -465,20 +465,53 @@ def get_items(warehouse, posting_date, posting_time, company):
...
+						"item_code": d[0],
+						"warehouse": d[2],
+						"package_tag": package_tag.package_tag,
+						"qty": package_tag.actual_qty,
+						"item_name": d[1],
+						"valuation_rate": package_tag.valuation_rate,
+						"current_qty": stock_bal[0],
+						"current_valuation_rate": stock_bal[1]
+					})
+			if batch_nos:
```
Comment (hrwX) [inline erpnext/stock/doctype/stock_reconciliation/stock_reconciliation.py:493]:
```suggestion
			elif batch_nos:
```

**Review by hrwX — CHANGES_REQUESTED** [review]

File: erpnext/stock/doctype/stock_reconciliation/stock_reconciliation.py (hunk 4 above)
Comment (MyuddinKhatri) [reply erpnext/stock/doctype/stock_reconciliation/stock_reconciliation.py:493]: kept this if purposely as there might be case where item has Batches and Package Tags both.

**Review by sahil28297 — APPROVED** [review]

**Pull request merged**